File: ADP.Surveys/ADP.Surveys.Shared/Evaluation/ExpressionSandbox/Parser.cs
namespace ShiftSoftware.ADP.Surveys.Shared.Evaluation.ExpressionSandbox;

/// <summary>
/// Hand-rolled recursive-descent parser. Grammar is described on <see cref="ExprNode"/>.
/// </summary>
internal sealed class Parser(List<Token> tokens)
{
    private readonly List<Token> tokens = tokens;
    private int pos;

    public ExprNode Parse()
    {
        var expr = ParseOr();
        Expect(TokenKind.EndOfInput);
        return expr;
    }

    private ExprNode ParseOr()
    {
        var left = ParseAnd();
        while (Match(TokenKind.Or))
            left = new BinaryOpNode("||", left, ParseAnd());
        return left;
    }

    private ExprNode ParseAnd()
    {
        var left = ParseEquality();
        while (Match(TokenKind.And))
            left = new BinaryOpNode("&&", left, ParseEquality());
        return left;
    }

    private ExprNode ParseEquality()
    {
        var left = ParseComparison();
        while (true)
        {
            string? op = Peek().Kind switch
            {
                TokenKind.Eq or TokenKind.StrictEq => "==",
                TokenKind.NotEq or TokenKind.StrictNotEq => "!=",
                _ => null,
            };
            if (op is null) break;
            Advance();
            left = new BinaryOpNode(op, left, ParseComparison());
        }
        return left;
    }

    private ExprNode ParseComparison()
    {
        var left = ParseUnary();
        while (true)
        {
            string? op = Peek().Kind switch
            {
                TokenKind.Lt => "<",
                TokenKind.Gt => ">",
                TokenKind.LtEq => "<=",
                TokenKind.GtEq => ">=",
                _ => null,
            };
            if (op is null) break;
            Advance();
            left = new BinaryOpNode(op, left, ParseUnary());
        }
        return left;
    }

    private ExprNode ParseUnary()
    {
        if (Match(TokenKind.Not))
            return new UnaryOpNode("!", ParseUnary());
        return ParsePrimary();
    }

    private ExprNode ParsePrimary()
    {
        var tok = Peek();
        switch (tok.Kind)
        {
            case TokenKind.Number:
            case TokenKind.String:
            case TokenKind.True:
            case TokenKind.False:
            case TokenKind.Null:
                Advance();
                return new LiteralNode(tok.Literal);

            case TokenKind.LParen:
                Advance();
                var inner = ParseOr();
                Expect(TokenKind.RParen);
                return inner;

            case TokenKind.LBracket:
                return ParseArray();

            case TokenKind.Identifier:
                return ParseIdentifierExpression();
        }

        throw new ExpressionSyntaxException($"unexpected token '{tok.Text}'.", tok.Position);
    }

    private ExprNode ParseArray()
    {
        Expect(TokenKind.LBracket);
        var items = new List<ExprNode>();
        if (Peek().Kind != TokenKind.RBracket)
        {
            items.Add(ParseOr());
            while (Match(TokenKind.Comma)) items.Add(ParseOr());
        }
        Expect(TokenKind.RBracket);
        return new ArrayNode(items);
    }

    private ExprNode ParseIdentifierExpression()
    {
        var ident = Advance();
        if (ident.Text == "answers")
            return ParseAnswersAccess(ident.Position);

        // Otherwise, must be a function call: ident(...).
        Expect(TokenKind.LParen);
        var args = new List<ExprNode>();
        if (Peek().Kind != TokenKind.RParen)
        {
            args.Add(ParseOr());
            while (Match(TokenKind.Comma)) args.Add(ParseOr());
        }
        Expect(TokenKind.RParen);
        return new CallNode(ident.Text, args);
    }

    private ExprNode ParseAnswersAccess(int position)
    {
        string key;
        if (Match(TokenKind.Dot))
        {
            var ident = Expect(TokenKind.Identifier);
            key = ident.Text;
        }
        else if (Match(TokenKind.LBracket))
        {
            var str = Expect(TokenKind.String);
            Expect(TokenKind.RBracket);
            key = (string)str.Literal!;
        }
        else
        {
            throw new ExpressionSyntaxException("'answers' must be followed by .key or ['key'].", position);
        }
        return new AnswersAccessNode(key);
    }

    private Token Peek() => tokens[pos];

    private Token Advance()
    {
        var t = tokens[pos];
        if (t.Kind != TokenKind.EndOfInput) pos++;
        return t;
    }

    private bool Match(TokenKind kind)
    {
        if (Peek().Kind != kind) return false;
        Advance();
        return true;
    }

    private Token Expect(TokenKind kind)
    {
        var t = Peek();
        if (t.Kind != kind)
            throw new ExpressionSyntaxException($"expected {kind}, got '{t.Text}'.", t.Position);
        Advance();
        return t;
    }
}

File: ADP.Surveys/renderer/packages/survey-sdk/src/expression-sandbox/parser.ts
/**
 * Hand-rolled recursive-descent parser. Grammar is described on ast.ts. Mirror of
 * `ADP.Surveys.Shared/Evaluation/ExpressionSandbox/Parser.cs`.
 */

import type { ExprNode, BinaryOperator } from './ast.js';
import { type Token, ExpressionSyntaxError } from './lexer.js';

export function parse(tokens: Token[]): ExprNode {
  let pos = 0;

  const peek = (): Token => {
    const t = tokens[pos];
    if (!t) throw new ExpressionSyntaxError('unexpected end of tokens.', 0);
    return t;
  };

  const advance = (): Token => {
    const t = peek();
    if (t.kind !== 'EndOfInput') pos++;
    return t;
  };

  const match = (kind: Token['kind']): boolean => {
    if (peek().kind !== kind) return false;
    advance();
    return true;
  };

  const expect = (kind: Token['kind']): Token => {
    const t = peek();
    if (t.kind !== kind) {
      throw new ExpressionSyntaxError(`expected ${kind}, got '${t.text}'.`, t.position);
    }
    advance();
    return t;
  };

  const parseOr = (): ExprNode => {
    let left = parseAnd();
    while (match('Or')) {
      left = { kind: 'BinaryOp', op: '||', left, right: parseAnd() };
    }
    return left;
  };

  const parseAnd = (): ExprNode => {
    let left = parseEquality();
    while (match('And')) {
      left = { kind: 'BinaryOp', op: '&&', left, right: parseEquality() };
    }
    return left;
  };

  const parseEquality = (): ExprNode => {
    let left = parseComparison();
    while (true) {
      const kind = peek().kind;
      let op: BinaryOperator | null = null;
      if (kind === 'Eq' || kind === 'StrictEq') op = '==';
      else if (kind === 'NotEq' || kind === 'StrictNotEq') op = '!=';
      if (op === null) break;
      advance();
      left = { kind: 'BinaryOp', op, left, right: parseComparison() };
    }
    return left;
  };

  const parseComparison = (): ExprNode => {
    let left = parseUnary();
    while (true) {
      const kind = peek().kind;
      let op: BinaryOperator | null = null;
      if (kind === 'Lt') op = '<';
      else if (kind === 'Gt') op = '>';
      else if (kind === 'LtEq') op = '<=';
      else if (kind === 'GtEq') op = '>=';
      if (op === null) break;
      advance();
      left = { kind: 'BinaryOp', op, left, right: parseUnary() };
    }
    return left;
  };

  const parseUnary = (): ExprNode => {
    if (match('Not')) return { kind: 'UnaryOp', op: '!', operand: parseUnary() };
    return parsePrimary();
  };

  const parseArray = (): ExprNode => {
    expect('LBracket');
    const items: ExprNode[] = [];
    if (peek().kind !== 'RBracket') {
      items.push(parseOr());
      while (match('Comma')) items.push(parseOr());
    }
    expect('RBracket');
    return { kind: 'Array', items };
  };

  const parseAnswersAccess = (position: number): ExprNode => {
    let key: string;
    if (match('Dot')) {
      const ident = expect('Identifier');
      key = ident.text;
    } else if (match('LBracket')) {
      const str = expect('String');
      expect('RBracket');
      key = str.literal as string;
    } else {
      throw new ExpressionSyntaxError(
        "'answers' must be followed by .key or ['key'].",
        position,
      );
    }
    return { kind: 'AnswersAccess', key };
  };

  const parseIdentifierExpression = (): ExprNode => {
    const ident = advance();
    if (ident.text === 'answers') return parseAnswersAccess(ident.position);

    // Otherwise must be a function call: ident(...)
    expect('LParen');
    const args: ExprNode[] = [];
    if (peek().kind !== 'RParen') {
      args.push(parseOr());
      while (match('Comma')) args.push(parseOr());
    }
    expect('RParen');
    return { kind: 'Call', name: ident.text, args };
  };

  const parsePrimary = (): ExprNode => {
    const tok = peek();
    switch (tok.kind) {
      case 'Number':
      case 'String':
      case 'True':
      case 'False':
      case 'Null':
        advance();
        return { kind: 'Literal', value: tok.literal };
      case 'LParen': {
        advance();
        const inner = parseOr();
        expect('RParen');
        return inner;
      }
      case 'LBracket':
        return parseArray();
      case 'Identifier':
        return parseIdentifierExpression();
      default:
        throw new ExpressionSyntaxError(`unexpected token '${tok.text}'.`, tok.position);
    }
  };

  const expr = parseOr();
  expect('EndOfInput');
  return expr;
}
